### PR TITLE
NOIDs should not be second-class citizens.

### PR DESCRIPTION
**Pull request by mjgiarlo**
Prior to Fedora 4, IDs in Hydra resembled Fedora identifiers ('namespace:id'), and in Sufia we had two opinions about this: 1) the namespace should never be shown to users; 2) the default ID in Fedora 3, a sequential integer, should be overridden by a configurable NOID. In Fedora 4, IDs no longer have a namespace, and so rather than continue to have two kinds of identifiers floating around the system, we have just one: the object identifier, which takes the form of a NOID. This PR eliminates nearly all references to noids in the code (since they're just object identifiers), and makes usage of NOIDs configurable. If you set Sufia.config.enable_noids to false, you will see Fedora 4's UUIDs instead of NOIDs.

### CODE REVIEW

File: sufia-models/app/services/sufia/indexing_service.rb [hunk 1]
```diff
@@ -1,15 +1,4 @@
 module Sufia
   class IndexingService < ActiveFedora::IndexingService
```
Comment (jcoyne) [inline sufia-models/app/services/sufia/indexing_service.rb]: Do we need this class anymore?

Comment (mjgiarlo) [inline sufia-models/app/services/sufia/indexing_service.rb]: Nope.

File: sufia-models/lib/generators/sufia/models/templates/config/sufia.rb [hunk 2]
```diff
@@ -105,8 +105,11 @@
   # If you have ffmpeg installed and want to transcode audio and video uncomment this line
   # config.enable_ffmpeg = true
 
-  # Specify the Fedora pid prefix:
-  # config.id_namespace = "sufia"
+  # To disable minting of NOIDs for files and collections in favor of Fedora UUIDs
+  # config.enable_noids = false
```
Comment (jcoyne) [inline sufia-models/lib/generators/sufia/models/templates/config/sufia.rb]: This should be `true` so it matches the default.

File: sufia-models/app/services/sufia/indexing_service.rb (hunk 1 above)
Comment (mjgiarlo) [inline sufia-models/app/services/sufia/indexing_service.rb]: Removed in latest push.

File: sufia-models/lib/generators/sufia/models/templates/config/sufia.rb (hunk 2 above)
Comment (mjgiarlo) [inline sufia-models/lib/generators/sufia/models/templates/config/sufia.rb]: Fixed in latest push.

File: sufia-models/app/services/sufia/generic_file_indexing_service.rb [hunk 3]
```diff
@@ -1,5 +1,5 @@
 module Sufia
-  class GenericFileIndexingService < IndexingService
+  class GenericFileIndexingService
```
Comment (jcoyne) [inline sufia-models/app/services/sufia/generic_file_indexing_service.rb]: Should extend `ActiveFedora::IndexingService`

File: sufia-models/lib/generators/sufia/models/templates/config/sufia.rb [hunk 4]
```diff
@@ -105,8 +105,11 @@
   # If you have ffmpeg installed and want to transcode audio and video uncomment this line
   # config.enable_ffmpeg = true
 
-  # Specify the Fedora pid prefix:
-  # config.id_namespace = "sufia"
+  # To disable minting of NOIDs for files and collections in favor of Fedora UUIDs
```
Comment (jcoyne) [inline sufia-models/lib/generators/sufia/models/templates/config/sufia.rb]: Change text to `Sufia uses NOIDs for files and collections instead of Fedora UUIDs`

File: sufia-models/app/services/sufia/generic_file_indexing_service.rb (hunk 3 above)
Comment (mjgiarlo) [inline sufia-models/app/services/sufia/generic_file_indexing_service.rb]: Derp.  Done.

File: sufia-models/lib/generators/sufia/models/templates/config/sufia.rb (hunk 4 above)
Comment (mjgiarlo) [inline sufia-models/lib/generators/sufia/models/templates/config/sufia.rb]: fixed with latest push.

**Comment by jcoyne** [comment]
:heavy_plus_sign: :100:  I'm :grinning:

File: app/helpers/sufia/sufia_helper_behavior.rb [hunk 5]
```diff
@@ -38,7 +38,7 @@ def sufia_thumbnail_tag(document, options)
       # file
       else
         path = if document.image? || document.pdf? || document.video? || document.office_document?
-          sufia.download_path document.noid, datastream_id: 'thumbnail'
+          sufia.download_path document, datastream_id: 'thumbnail'
         elsif document.audio?
```
Comment (cam156) [inline app/helpers/sufia/sufia_helper_behavior.rb]: Should this be document.id?

Comment (cam156) [inline app/helpers/sufia/sufia_helper_behavior.rb]: NM I see you are consistently not using id in path helpers

File: app/helpers/generic_file_helper.rb [hunk 6]
```diff
@@ -18,7 +18,7 @@ def render_download_icon title = nil
   end
 
   def render_download_link text = nil
-    link_to (text || "Download"), sufia.download_path(@generic_file.noid), { id: "file_download", target: "_new", data: { label: @generic_file.id } }
+    link_to (text || "Download"), sufia.download_path(@generic_file.id), { id: "file_download", target: "_new", data: { label: @generic_file.id } }
   end
```
Comment (cam156) [inline app/helpers/generic_file_helper.rb]: This should not have id to match up with you other uses of path helpers

File: app/views/collections/_sort_and_per_page.html.erb [hunk 7]
```diff
@@ -12,7 +12,7 @@
    <div class="sort-toggle">
      <%# kind of hacky way to get this to work on catalog and folder controllers.  May be able to simple do {action: "index"} but I'm not sure -%>
      <% unless @response.response['numFound'] < 2 %>
-        <%= form_tag collections.collection_path(@collection.noid), method: :get, class: 'per_page form-inline' do %>
+        <%= form_tag collections.collection_path(@collection.id), method: :get, class: 'per_page form-inline' do %>
           <%= label_tag(:sort, "<span>Sort By:</span>".html_safe) %>
```
Comment (cam156) [inline app/views/collections/_sort_and_per_page.html.erb]: I think just passing the collection here to match with other paths might be good

File: app/views/homepage/_recent_document.html.erb [hunk 8]
```diff
@@ -9,7 +9,7 @@
       <% end %>
       <td>
         <h3>
-          <span class="sr-only">Title</span><%= link_to truncate(recent_document.title_or_label, length: 28, separator: ' '), sufia.generic_file_path(recent_document.noid)  %>
+          <span class="sr-only">Title</span><%= link_to truncate(recent_document.title_or_label, length: 28, separator: ' '), sufia.generic_file_path(recent_document.id)  %>
           <% if display_access %>
```
Comment (cam156) [inline app/views/homepage/_recent_document.html.erb]: Here too...

File: app/views/homepage/_recent_document.html.erb [hunk 9]
```diff
@@ -22,7 +22,7 @@
         </h3>
         <p>
           <span class="sr-only">Filename</span><%= link_to truncate(recent_document.label, length: 25),
-              sufia.generic_file_path(recent_document.noid), title: recent_document.label %><br />
+              sufia.generic_file_path(recent_document.id), title: recent_document.label %><br />
```
Comment (cam156) [inline app/views/homepage/_recent_document.html.erb]: And Here

File: spec/models/solr_document_spec.rb [hunk 10]
```diff
@@ -11,15 +11,6 @@
     end
   end
 
-  describe "to_param" do
```
Comment (cam156) [inline spec/models/solr_document_spec.rb]: Should this really be removed completely or should it check to see that id was assigned

File: sufia-models/app/jobs/resolrize_job.rb [hunk 11]
```diff
@@ -4,13 +4,6 @@ def queue_name
   end
 
   def run
-    require 'active_fedora/version'
```
Comment (cam156) [inline sufia-models/app/jobs/resolrize_job.rb]: Is this really related to this commit?  Seems unrelated to me

File: app/helpers/sufia/sufia_helper_behavior.rb (hunk 5 above)
Comment (mjgiarlo) [inline app/helpers/sufia/sufia_helper_behavior.rb]: You're right, I'm relying on `#to_param` magic rather than explicitly specifying the `#id` method. (And whether you use UUIDs or NOIDs, `#to_param` will return the object's identifier.)

File: app/helpers/generic_file_helper.rb (hunk 6 above)
Comment (mjgiarlo) [inline app/helpers/generic_file_helper.rb]: Good catch!

File: app/views/collections/_sort_and_per_page.html.erb (hunk 7 above)
Comment (mjgiarlo) [inline app/views/collections/_sort_and_per_page.html.erb]: Agreed.

File: app/views/homepage/_recent_document.html.erb (hunk 8 above)
Comment (mjgiarlo) [inline app/views/homepage/_recent_document.html.erb]: Thx!

File: app/views/homepage/_recent_document.html.erb (hunk 9 above)
Comment (mjgiarlo) [inline app/views/homepage/_recent_document.html.erb]: I detect a trend.

File: spec/models/solr_document_spec.rb (hunk 10 above)
Comment (mjgiarlo) [inline spec/models/solr_document_spec.rb]: Nope, it shouldn't be -- this was from an earlier run through the code when I excised that completely. I'll add the test back.

File: sufia-models/app/jobs/resolrize_job.rb (hunk 11 above)
Comment (mjgiarlo) [inline sufia-models/app/jobs/resolrize_job.rb]: resolrize_job was basically broken in its current form because it reindexed with an (F3-style) assumption that identifiers start with PIDs. This change makes reindexing work again under F4.

File: app/helpers/sufia/sufia_helper_behavior.rb [hunk 12]
```diff
@@ -154,8 +154,8 @@ def search_form_action
     end
 
     def render_visibility_link document
-      link_to render_visibility_label(document), sufia.edit_generic_file_path(document.noid, {anchor: "permissions_display"}),
-        id: "permission_"+document.noid, class: "visibility-link"
+      link_to render_visibility_label(document), sufia.edit_generic_file_path(document.id, {anchor: "permissions_display"}),
```
Comment (awead) [inline app/helpers/sufia/sufia_helper_behavior.rb]: but your'e using `id` in this path. Should it just be `document` as above in @cam156 comment?

File: lib/generators/sufia/templates/catalog_controller.rb [hunk 13]
```diff
@@ -124,7 +124,7 @@ def self.modified_field
       all_names = config.show_fields.values.map{|val| val.field}.join(" ")
       title_name = solr_name("title", :stored_searchable)
       field.solr_parameters = {
-        qf: "#{all_names} noid_tsi file_format_tesim all_text_timv",
+        qf: "#{all_names} file_format_tesim all_text_timv",
```
Comment (hectorcorrea) [inline lib/generators/sufia/templates/catalog_controller.rb]: I assume it is intentional here not to pass "id" in the "qf" parameter. Right?

File: app/views/collections/_show_document_list_row.html.erb [hunk 14]
```diff
@@ -1,10 +1,10 @@
-<% noid = document.noid %>
-<tr id="document_<%= noid %>" class="<%= cycle("", "zebra") %>">
+<% id = document.id %>
```
Comment (awead) [inline app/views/collections/_show_document_list_row.html.erb]: Couldn't this just be `document` and avoid the extra variable setting?

File: spec/services/noid_spec.rb [hunk 15]
```diff
@@ -1,16 +1,9 @@
...
-      let(:id) { 'abc123' }
-      it { is_expected.to eq 'sufia:abc123' }
-    end
-    context "when the passed in pid has a namespace" do
-      let(:id) { 'ksl:abc123' }
-      it { is_expected.to eq 'ksl:abc123' }
-    end
+  describe "#treeify" do
+    subject { Sufia::Noid.treeify(id) }
+    let(:id) { 'abc123def45' }
```
Comment (awead) [inline spec/services/noid_spec.rb]: Is there a min or max on the number of characters in the noid that Sufia will "treeify" ?

File: app/views/collections/_show_document_list_row.html.erb (hunk 14 above)
Comment (jcoyne) [inline app/views/collections/_show_document_list_row.html.erb]: @awead No, because it's used to make a string for the `id` attribute.

File: sufia-models/lib/generators/sufia/models/templates/config/sufia.rb [hunk 16]
```diff
@@ -105,8 +105,11 @@
   # If you have ffmpeg installed and want to transcode audio and video uncomment this line
   # config.enable_ffmpeg = true
 
-  # Specify the Fedora pid prefix:
-  # config.id_namespace = "sufia"
+  # Sufia uses NOIDs for files and collections instead of Fedora UUIDs
```
Comment (hectorcorrea) [inline sufia-models/lib/generators/sufia/models/templates/config/sufia.rb]: It might be good to clarify that NOIDs are in the form "nnnn" where as Fedora UUIDs are in the form "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx". Not sure how much common knowledge this is.

File: sufia-models/app/jobs/resolrize_job.rb (hunk 11 above)
Comment (cam156) [inline sufia-models/app/jobs/resolrize_job.rb]: :ok_hand:

File: lib/generators/sufia/templates/catalog_controller.rb (hunk 13 above)
Comment (mjgiarlo) [inline lib/generators/sufia/templates/catalog_controller.rb]: Yep! The ID is automatically included in the solr document.

File: app/views/collections/_show_document_list_row.html.erb (hunk 14 above)
Comment (mjgiarlo) [inline app/views/collections/_show_document_list_row.html.erb]: True, though I did miss a bunch of to_param opportunities in this file, so thanks for raising this.

File: spec/services/noid_spec.rb (hunk 15 above)
Comment (mjgiarlo) [inline spec/services/noid_spec.rb]: https://github.com/projecthydra/sufia/blob/master/sufia-models/app/services/sufia/noid.rb#L27

File: sufia-models/lib/generators/sufia/models/templates/config/sufia.rb (hunk 16 above)
Comment (mjgiarlo) [inline sufia-models/lib/generators/sufia/models/templates/config/sufia.rb]: done!

**Comment by mjgiarlo** [comment]
Testing latest round of changes, then pushing.
